scarb/src/ops/lint.rs

```diff
@@ -0,0 +1,237 @@
+    },
+    core::TargetKind,
+    ops,
+};
+use annotate_snippets::Renderer;
+use anyhow::{anyhow, Result};
+use cairo_lang_defs::db::DefsGroup;
+use cairo_lang_diagnostics::{DiagnosticEntry, Maybe};
+use cairo_lang_filesystem::db::FilesGroup;
+use cairo_lang_filesystem::ids::{CrateLongId, FileId};
+use cairo_lang_semantic::db::SemanticGroup;
+use cairo_lang_semantic::diagnostic::SemanticDiagnosticKind;
+use cairo_lang_syntax::node::SyntaxNode;
+use cairo_lang_utils::Upcast;
+use cairo_lint_core::{
+    diagnostics::format_diagnostic,
+    fix::{apply_import_fixes, collect_unused_imports, fix_semantic_diagnostic, Fix, ImportFix},
+    plugin::{cairo_lint_plugin_suite, diagnostic_kind_from_message, CairoLintKind},
+};
+use scarb_ui::components::Status;
+use smol_str::SmolStr;
+use std::cmp::Reverse;
+use std::collections::HashMap;
+
+use crate::core::{Package, Workspace};
+
+use super::{CompilationUnitsOpts, FeaturesOpts, FeaturesSelector};
+
+pub struct LintOptions {
+    pub packages: Vec<Package>,
+    pub test: bool,
+    pub fix: bool,
+    pub ignore_cairo_version: bool,
+}
+
+#[tracing::instrument(skip_all, level = "debug")]
+pub fn lint(opts: LintOptions, ws: &Workspace<'_>) -> Result<()> {
+    let feature_opts = FeaturesOpts {
+        features: FeaturesSelector::AllFeatures,
+        no_default_features: true,
+    };
+
+    let resolve = ops::resolve_workspace(ws)?;
+
+    let compilation_units = ops::generate_compilation_units(
+        &resolve,
+        &feature_opts,
+        ws,
+        CompilationUnitsOpts {
+            ignore_cairo_version: opts.ignore_cairo_version,
+            load_prebuilt_macros: true,
+        },
+    )?;
+
+    for package in opts.packages {
+        let package_compilation_units = if opts.test {
+            compilation_units
+                .iter()
+                // TODO: Test this (if the test packages are also checked)
+                .filter(|compilation_unit| {
+                    let is_main_component = compilation_unit.main_package_id() == package.id;
+                    let has_test_components =
+                        compilation_unit.components().iter().any(|component| {
+                            component.target_kind() == TargetKind::TEST
+                                && component.package.id == package.id
+                        });
+                    let is_integration_test_package =
+                        compilation_unit.main_package_id().name.to_string()
+                            == format!("{}_integrationtest", package.id.name)
+                            && compilation_unit.main_package_id().version == package.id.version;
+                    is_main_component || has_test_components || is_integration_test_package
+                })
+                .collect::<Vec<_>>()
+        } else {
+            vec![compilation_units
+                .iter()
+                .find(|compilation_unit| compilation_unit.main_package_id() == package.id)
```

Comment:
Note this would also check package units tests - see my previous comment.

Reply:
Filtered out CU that has `"target": "test"` in cfgset 👍